Refactor context function, add supporting tests

capjamesg opened the pull request:
Per issue #2, I have refactored the reply context function.

The main changes are:

1. A custom data class is used for the reply context instead of a dictionary.
2. New exceptions are defined.
3. The code has been slightly reorganised.

I have written one example test case which uses my own site as an example. The test case passes. But, I intend to write more (and could use with help on writing more) so that the function is as robust as possible.

capjamesg committed:
refactor context function, add supporting tests

jamesvandyne commented on src/indieweb_utils/replies/context.py:
```diff
@@ -29,298 +40,316 @@ def get_reply_context(url, twitter_bearer_token=None):
-            page_content = requests.get(url, timeout=10, verify=False, headers=http_headers)
-        except:
-            return True, {"error": "Page content could not be retrieved."}
+    if not url.startswith("https://") and not url.startswith("http://"):
```
Since we have the parsed url already split, we could simplify this.

```suggestion
    if parsed_url.scheme not in ["http", https"]:
```

jamesvandyne commented on src/indieweb_utils/replies/context.py:
```diff
@@ -29,298 +40,316 @@ def get_reply_context(url, twitter_bearer_token=None):
-        except:
-            return True, {"error": "Page content could not be retrieved."}
+    if not url.startswith("https://") and not url.startswith("http://"):
+        return return_object
```
If it's not an http url, we don't support it, right? Rather than return an empty class object, we should raise an error.

```python
raise UnsupportedScheme(f"{parsed_url.scheme} is not supported")
```

jamesvandyne commented on src/indieweb_utils/replies/context.py:
```diff
@@ -29,298 +40,316 @@ def get_reply_context(url, twitter_bearer_token=None):
 
-        site_supports_webmention, _ = discover_webmention_endpoint(url)
+    if page_content.status_code != 200:
+        raise ReplyContextRetrievalError("Page did not return a 200 response.")
```
It would be helpful to put the actual returned status code in the error message.

🤔  I wonder if we there aren't any special cases we'd want to handle (as a user of the library) a particular status code _different_. I'm thinking mostly like a 500 or 404?  Just a thought, no need to make changes.

jamesvandyne commented on src/indieweb_utils/replies/context.py:
```diff
@@ -29,298 +40,316 @@ def get_reply_context(url, twitter_bearer_token=None):
-            author_url = None
-            author_name = None
-            author_image = None
+    domain = url.replace("https://", "").replace("http://", "").split("/")[0]
```
Could this use our parsed url?

jamesvandyne commented on src/indieweb_utils/replies/context.py:
```diff
@@ -10,14 +11,24 @@
+    video: str
+    post_html: str
+    post_text: str
     author_name: str
```
Could an article have multiple authors? I think we should split author into its own dataclass, then we can use it in our reply context. 

```
authors: List[PostAuthor]
```

If we're unable to find an author, this would just be an empty list, so it would be easy to use as a user of the library, too, just literate over it.

capjamesg committed:
create PostAuthor dataclass, add unsupported scheme exception, make minor changes

capjamesg committed:
Merge branch 'main' into context_refactor

capjamesg commented:
@jamesvandyne I have made changes per your comments and have also added a new PostAuthor dataclass. I updated our test case so it looks out for author attributes in the dataclass instead of in separate attributes.

jamesvandyne commented on src/indieweb_utils/replies/context.py:
```diff
@@ -29,298 +48,306 @@ def get_reply_context(url, twitter_bearer_token=None):
+            post_body = ""
 
-            return True, h_entry, site_supports_webmention
+        return_object.name = ""
+        return_object.post_url = url
+        return_object.post_text = post_body
+        return_object.post_html = post_body
+        return_object.authors = [PostAuthor(url=url, name=author_name, photo=author_image)]
+        return_object.photo = ""
+        return_object.video = ""
```
When using a dataclass, you should instantiate it instead of setting attributes on an un-instantiated class.

```python
return ReplyContext(
    name = ""
    post_url=post_url
    post_body=post_body
   ...
)
```

capjamesg committed:
divide code into new functions, address feedback on dataclasses

capjamesg commented:
@jamesvandyne I have refactored the code in the context.py file so that it is easier to read. This should also help with linter warnings about the function being too complex. I have made the change to the dataclass usage that you requested. Please review my edits.

jamesvandyne commented on src/indieweb_utils/replies/context.py:
```diff
@@ -7,320 +9,302 @@
+    author_image = ""
+
+    if h_entry["properties"].get("author"):
+        if type(h_entry["properties"]["author"][0]) == dict and h_entry["properties"]["author"][0].get(
```
It's better to use `isinstance` when checking the type of an object. e.g. 
```python
isinstance(h_entry["properties"]["author"][0], dict)
```

capjamesg replied:
Good suggestion!

jamesvandyne commented on src/indieweb_utils/replies/context.py:
```diff
@@ -7,320 +9,302 @@
-
-            if h_entry["properties"].get("photo"):
-                post_photo_url = canonicalize_url(h_entry["properties"]["photo"][0], domain, url)
+                author_name = ""
```
Since author_name is defined at the top of the function, probably don't need to set it blank here, too?

jamesvandyne commented on src/indieweb_utils/replies/context.py:
```diff
@@ -7,320 +9,302 @@
+        else:
+            author_image = ""
+
+        post_body = " ".join(post_body.split(" ")[:75]) + " ..."
```
75 should probably be passed in as a kwarg to this function so it can be overridden, but it's ok as is for now.

jamesvandyne commented on tests/test_indieweb_utils.py:
```diff
@@ -54,3 +54,22 @@ def test_webpage_feed_discovery(self):
+    # TODO: Add test cases for Twitter
+    # TODO: Add test cases for other websites
+    # TODO: Add test cases with photo / video responses
+    def test_reply_context_discovery(self):
```
I imagine you're planning to use responses to prevent network requests in this test, right?

jamesvandyne reviewed (approved):
LGTM! Good work splitting up the concerns into smaller functions.

capjamesg committed:
address feedback from pr #15 thread

capjamesg commented:
@jamesvandyne Thank you for your review! I do plan to write mock network requests for these changes. However, i have to address a few merge conflicts before proceeding.

capjamesg commented:
I have addressed the merge conflict with the test suite. I'll include the mock network requests in a separate PR.
